chromium_src/chrome/browser/ui/webui/tab_search/tab_search_page_handler.cc

```diff
@@ -0,0 +1,255 @@
+  ai_chat::AIChatService* ai_chat_service =
+      ai_chat::AIChatServiceFactory::GetForBrowserContext(profile);
+  CHECK(ai_chat_service);
+
+  tab_search::mojom::ErrorPtr error = tab_search::mojom::Error::New();
+  if (api_error == ai_chat::mojom::APIError::RateLimitReached) {
+    bool is_premium = ai_chat_service->IsPremiumStatus();
+    error->message =
+        is_premium
+            ? l10n_util::GetStringUTF8(IDS_CHAT_UI_ERROR_RATE_LIMIT)
+            : l10n_util::GetStringUTF8(IDS_CHAT_UI_RATE_LIMIT_REACHED_DESC);
+    error->rate_limited_info =
+        tab_search::mojom::RateLimitedInfo::New(is_premium);
+  } else if (api_error == ai_chat::mojom::APIError::ConnectionIssue) {
+    error->message = l10n_util::GetStringUTF8(IDS_CHAT_UI_ERROR_NETWORK);
+  } else {
+    error->message = l10n_util::GetStringUTF8(IDS_CHAT_UI_ERROR_INTERNAL);
+  }
+
+  return error;
+}
+
+void TabSearchPageHandler::GetSuggestedTopics(
+    GetSuggestedTopicsCallback callback) {
+  auto* engine = GetAIEngineForTabOrganization();
+
+  std::vector<ai_chat::Tab> tabs = GetTabsForAIEngine();
+  engine->GetSuggestedTopics(
+      tabs,
+      base::BindOnce(&TabSearchPageHandler::OnGetSuggestedTopics,
+                     weak_ptr_factory_.GetWeakPtr(), std::move(callback)));
+}
+
+void TabSearchPageHandler::OnGetSuggestedTopics(
+    GetSuggestedTopicsCallback callback,
+    base::expected<std::vector<std::string>, ai_chat::mojom::APIError> result) {
+  if (!result.has_value()) {
+    std::move(callback).Run({}, GetError(result.error()));
+    return;
+  }
+
+  std::move(callback).Run(*result, nullptr);
+}
+
+void TabSearchPageHandler::GetFocusTabs(const std::string& topic,
+                                        GetFocusTabsCallback callback) {
+  auto* engine = GetAIEngineForTabOrganization();
+  focus_tabs_info_.clear();
+
+  std::vector<ai_chat::Tab> tabs = GetTabsForAIEngine();
+  engine->GetFocusTabs(
+      tabs, topic,
+      base::BindOnce(&TabSearchPageHandler::OnGetFocusTabs,
+                     weak_ptr_factory_.GetWeakPtr(), std::move(callback)));
+}
+
+void TabSearchPageHandler::OnGetFocusTabs(
+    GetFocusTabsCallback callback,
+    base::expected<std::vector<std::string>, ai_chat::mojom::APIError> result) {
+  if (!result.has_value()) {
+    std::move(callback).Run(false, GetError(result.error()));
+    return;
+  }
+
+  // Move all tabs from normal browser window to a new window.
+  std::vector<TabDetails> tab_details_before_move;
+  for (const auto& tab_id_str : *result) {
+    int tab_id = 0;
+    if (!base::StringToInt(tab_id_str, &tab_id)) {
+      continue;
+    }
+
+    std::optional<TabDetails> details = GetTabDetails(tab_id);
+    if (!details) {
+      continue;
+    }
+
+    // Store all details before move to preserve the index.
+    tab_details_before_move.push_back(*details);
+    // Store old window ID (session ID), tab ID, and tab strip index for
+    // undo.
+    focus_tabs_info_[details->browser->session_id()].push_back(
```

Comment:
I wonder whether
`previous_tabs_for_session_` would be a better name, as its just what we undo to

Reply:
rename to original_tabs_info_by_window_ and update the comment in https://github.com/brave/brave-core/pull/28103/commits/1d61f28443c29de1b164f4bc4267f82c8a529dca